extensions/2.0/Khronos/KHR_materials_specular/README.md

```diff
@@ -0,0 +1,110 @@
+
+| |Type|Description|Required|
+|-|----|-----------|--------|
+| **specularFactor** | `number` | The strength of the specular reflection. | No, default: `1.0`|
+| **specularTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo) | A grayscale texture that defines the specular factor. Will be multiplied by specularFactor. | No |
+| **specularColorFactor** | `number[3]` | The F0 color of the specular reflection (RGB). | No, default: `[1.0, 1.0, 1.0]`|
+| **specularColorTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo) | A 1-channel luminance or 3-channel RGB texture that defines the F0 color of the specular reflection. Will be multiplied by specularColorFactor. | No |
+
+### Conversions
+
+Material models that define f0 in terms of reflectance at normal incidence (with 0 meaning 0%, 0.5 meaning 4%, and 1 meaning 8%) can be converted to `specularColorFactor` and `specularColorTexture` as follows:
+
+```
+specularColorFactor = 2 * reflectanceFactor
+specularColorTexture = reflectanceTexture
+```
+
+Alternatively, the index of refraction defined in `KHR_materials_ior` can be set to the value 1.788789, resulting in a f0 scale factor of 0.08.
+
+In addition, it is possible to convert specular-glossiness materials (`KHR_materials_pbrSpecularGlossiness`) to metallic-roughness by combining `KHR_materials_specular` and `KHR_materials_ior`. The conversion is lossless and makes effects like sheen (`KHR_materials_sheen`) and clearcoat (`KHR_materials_clearcoat`) available to specular-glossiness materials.
+
+```
+ior = inf
+specularColorFactor = specularFactor_specgloss
+specularColorTexture = specularGlossinessTexture_specgloss[0:2]
+```
+
+The conversion is not compatible with transmissive materials defined via `KHR_materials_transmission` and `KHR_materials_volume`.
+
+## Implementation
+
+The specular factor scales the microfacet BRDF in the dielectric BRDF. It also affects the diffuse BRDF; the less energy is reflected by the microfacet BRDF, the more can be shifted to the diffuse BRDF.
+
+The specular color changes the F0 color of the Fresnel that is multiplied to the microfacet BRDF. The color at grazing angles (F90) is not changed.
+
+```
+dielectric_brdf =
+  fresnel_mix(
+    diffuse_brdf(baseColor),
+    microfacet_brdf(roughness^2),
+    f0 = 0.04 * specular * specularColor,
+    f90 = specular)
+```
+
+Based on [Appendix B](/specification/2.0/README.md#appendix-b-brdf-implementation), the modified `fresnel_mix` operation is defined as follows:
+
+```
+fresnel_mix(base, layer, ior) = base * (1 - fr(ior)) + layer * fr(ior)
+fr(ior) = f0 + (f90 - f0) * (1 - cos)^5
+f0 = 0.04 * specular * specularColor
+f90 = specular
+```
+
+If `KHR_materials_ior` is used in combination with `KHR_materials_specular`, the constant `0.04` is replaced by the value computed from the IOR:
+
+```
+f0 = ((ior - outside_ior) / (ior + outside_ior))^2 * specular * specularColor
+f90 = specular
+```
```

Comment:
We can avoid the specular factor texture read if we premultiply it with the specular color.
In order to do this we need to change the calculations a bit:

```
dielectric_brdf =
  fresnel_mix(
    diffuse_brdf(baseColor),
    microfacet_brdf(roughness^2),
    specularColor)
```

`fresnel_mix` would be defined as:

```
fresnel_mix(base, layer, ior) = base * (1 - fr(ior) * specularColor) + layer * fr(ior) * specularColor
fr(ior) = f0 + (f90 - f0) * (1 - cos)^5
f0 = 0.04
f90 = 1.0
```

And the version using IOR becomes:

```
f0 = ((ior - outside_ior) / (ior + outside_ior))^2
f90 = 1.0
```

This changes the attenuation of the base layer to include the specular color as well, which is actually more physically plausible.
This way we decouple the specular color from the Fresnel calculation and we are free to change the Fresnel implementation if we want.
Additionally we avoid modifying `f90`, which should always be 1.
This is actually what V-Ray does and alSurface does a monochrome version of it:

```
fresnel_mix(base, layer, ior) = base * (1 - fr(ior) * maxComponent(specularColor)) + layer * fr(ior) * maxComponent(specularColor))
```
where `maxComponent` returns the maximum of the R,G and B components.

Reply:
There are a couple of reasons why `specular_color` is not used as a directional-independent multiplier to the specular BSDF in this extension. We thought about it a lot when we designed the [Enterprise PBR Material](https://dassaultsystemes-technology.github.io/EnterprisePBRShadingModel/spec-2021x.md.html), which is the foundation for #1718, #1719, and a few other extensions proposed for glTF PBR Next (click [here](https://github.com/DassaultSystemes-Technology/glTF/tree/pbr-next-extensions/specification/2.0#user-content-extensions) for an overview). I think our guidelines and trade-offs apply equally well to glTF, therefore I'd like to explain them now.

*Compatibility to other material models*

As far as I know there are two competing camps wrt. specular color.
1. The Disney Principled BRDF uses a single-channel parameter called specularTint to blend F0 color of the Fresnel to the base color. This also applies to its derivatives, like the Blender Principled BSDF.
2. Others like Standard Surface and V-Ray use a color parameter to tint the specular BSDF independent of the direction.

Moreover, there is the specular-glossiness workflow and the glTF extension KHR_materials_pbrSpecularGlossiness. In this case, the specular color is also used as F0 color. Therefore, although not directly comparable (the workflow is different) I would put them into the first camp, together with Disney BRDF.

Our proposal/Enterprise PBR now is a compromise that can represent all these material models to a certain extend.
* There is a lossless conversion from Disney BRDF: Disney parameters `specular` and `specularTint` can be mapped to `specularColor`. `specular` alone can be mapped without preprocessing (mentioned in the "Conversion" section in the extension), `specularTint` needs preprocessing at export time.
* There is a lossless conversion from KHR_materials_pbrSpecularGlossiness. It doesn't need preprocessing of textures. I think this will help a lot in adopting the new material model, because a) engines need to implement only one Ubershader and b) artists can slowly migrate their materials to the new model (for example if they want a clear-coat for an existing spec-gloss model, they can transition to glTF PBR Next without rebuilding the whole material).
* There is a lossy conversion from Standard Surface and V-Ray: the intensity of the directional-independent specular color can map to the `specular` parameter, the color to the `specularColor` parameter. It's lossy, but I think it is acceptable. Given the fact that glTF PBR Next has a lot less parameters than Standard Surface/V-Ray to make it rasterizer-friendly (Mobile, WebGL, ...), there will be some loss anyway, from other areas as well.

*IOR and number of texture fetches*

> We can avoid the specular factor texture read if we premultiply it with the specular color.

I guess in your proposal the IOR is texturable, right? If yes, then we need a 4-channel texture: IOR (float) + specular color (float3). This is the same amount as in #1718 (KHR_materials_ior) and #1719.

The IOR in #1718 is not texturable, because it conceptually describes the refractive index of interface _and_ the volume beneath the surface, not only the refractive index of the interface. I think this distinction is important, because only a uniform volume IOR ensures reciprocity (think of light tracing and/or nested dielectrics). Of course, artists want to use a texture to modulate the specular contribution, and, therefore, we provide them the parameters `specular` and `specularColor`. These parameters don't mess with the volume IOR, so they are reciprocal. I think we should be very strict about reciprocity (and energy conservation). The normative part of glTF must not require something that violates these principles. This way we ensure consistency across a large variety of rendering algorithms, from rasterizers to bidirectional path tracing, and be somewhat future-proof.

To conclude, with #1718 and #1719, we also need only a 4-channel texture: specular (float) + specular color (float3). The uniform IOR is not textured.

In addition, there was some concerns regarding an IOR texture (see [this comment](https://github.com/KhronosGroup/glTF/pull/1718#discussion_r381957867)), because of the additional overhead it introduces to convert from 1/ior (stored in the texture) to F0 (used in Fresnel).

*Decoupling Fresnel and specular color*

> This way we decouple the specular color from the Fresnel calculation and we are free to change the Fresnel implementation if we want.

This is a good point, because probably some renderers want to use the full Fresnel equations, not the Schlick approximation. The specular color as it is defined in the extension works best with the Schlick approximation. However, I would consider this an implementation detail, therefore it is part of the non-normative section in the extension (not yet marked as such). It is possible to deal with it in combination with the full Fresnel equation, and I am happy to include this solution in the extension.

For example, I think it is possible to implement `fresnel_mix` as follows (untested, should work):

```
fresnel_mix(base, layer, ior) = base * (1 - specular * colored_fresnel(ior, specularColor)) + layer * specular * colored_fresnel(ior, specularColor)
colored_fresnel(ior, specularColor) = mix(specularColor * f0, 1, (fr(ior) - f0)/(1 - f0))
fr(ior) = ... full fresnel with ior and outside_ior ...
f0 = ((ior - outside_ior) / (ior + outside_ior))^2
```

In case `fr(ior)` is implemented with Schlick, it simplifies to

```
fresnel_mix(base, layer, ior) = base * (1 - specular * fr(ior, specularColor)) + layer * specular * fr(ior, specularColor)
fr(ior, specularColor) = f0 + (1 - f0) * (1 - cos)^5
f0 = ((ior - outside_ior) / (ior + outside_ior))^2 * specularColor
```